Convert geometry generation to Layer Builder

seanlam-ece opened the pull request:
https://optics.ansys.com/hc/en-us/articles/360034382394-Layer-builder-Simulation-object

Steps:
1. Convert LayerStack to Lumerical process file (.lbr)
2. Import process file and component GDS into Lumerical session


Notes / Known Issues:
- LayerStack conversion includes dopants, but this will create errors for Lumerical version 2021 and prior. Ensure that the layer stack does not have dopants if working with Lumerical 2021 and prior.
- Process file is saved as "process.lbr" in specified parent directory. Else, defaults to script directory.
- Pytest passes if either "passives only" or "passives + dopants" process files pass. This is done this way to pass when Lumerical 2021 is used since it only supports passives only, which is in its spec.

simbilod committed:
draft get_component_with_local_layers

dependabot[bot] committed:
Bump actions/upload-artifact from 3 to 4

Bumps [actions/upload-artifact](https://github.com/actions/upload-artifact) from 3 to 4.
- [Release notes](https://github.com/actions/upload-artifact/releases)
- [Commits](https://github.com/actions/upload-artifact/compare/v3...v4)

---
updated-dependencies:
- dependency-name: actions/upload-artifact
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump actions/upload-pages-artifact from 2 to 3

Bumps [actions/upload-pages-artifact](https://github.com/actions/upload-pages-artifact) from 2 to 3.
- [Release notes](https://github.com/actions/upload-pages-artifact/releases)
- [Commits](https://github.com/actions/upload-pages-artifact/compare/v2...v3)

---
updated-dependencies:
- dependency-name: actions/upload-pages-artifact
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump actions/setup-python from 4 to 5

Bumps [actions/setup-python](https://github.com/actions/setup-python) from 4 to 5.
- [Release notes](https://github.com/actions/setup-python/releases)
- [Commits](https://github.com/actions/setup-python/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/setup-python
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump actions/stale from 8 to 9

Bumps [actions/stale](https://github.com/actions/stale) from 8 to 9.
- [Release notes](https://github.com/actions/stale/releases)
- [Changelog](https://github.com/actions/stale/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/stale/compare/v8...v9)

---
updated-dependencies:
- dependency-name: actions/stale
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump actions/deploy-pages from 2 to 4

Bumps [actions/deploy-pages](https://github.com/actions/deploy-pages) from 2 to 4.
- [Release notes](https://github.com/actions/deploy-pages/releases)
- [Commits](https://github.com/actions/deploy-pages/compare/v2...v4)

---
updated-dependencies:
- dependency-name: actions/deploy-pages
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.1.8 → v0.1.11](https://github.com/astral-sh/ruff-pre-commit/compare/v0.1.8...v0.1.11)
- [github.com/macisamuele/language-formatters-pre-commit-hooks: v2.11.0 → v2.12.0](https://github.com/macisamuele/language-formatters-pre-commit-hooks/compare/v2.11.0...v2.12.0)

joamatab committed:
fix path_length_analysis

joamatab committed:
Merge pull request #293 from gdsfactory/fix_path_length_analysis

fix path_length_analysis

dependabot[bot] committed:
Bump tidy3d from 2.5.0 to 2.5.1

Bumps [tidy3d](https://github.com/flexcompute/tidy3d) from 2.5.0 to 2.5.1.
- [Release notes](https://github.com/flexcompute/tidy3d/releases)
- [Changelog](https://github.com/flexcompute/tidy3d/blob/develop/CHANGELOG.md)
- [Commits](https://github.com/flexcompute/tidy3d/compare/v2.5.0...v2.5.1)

---
updated-dependencies:
- dependency-name: tidy3d
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

joamatab committed:
Bump to 0.9.11

joamatab committed:
update changelog

joamatab committed:
Merge pull request #294 from gdsfactory/dependabot/pip/tidy3d-2.5.1

Bump tidy3d from 2.5.0 to 2.5.1

joamatab committed:
Merge pull request #283 from gdsfactory/pre-commit-ci-update-config

[pre-commit.ci] pre-commit autoupdate

joamatab committed:
Merge branch 'main' into dependabot/github_actions/actions/upload-artifact-4

joamatab committed:
Merge branch 'main' into dependabot/github_actions/actions/stale-9

joamatab committed:
Merge branch 'main' into dependabot/github_actions/actions/setup-python-5

joamatab committed:
Merge branch 'main' into dependabot/github_actions/actions/deploy-pages-4

sequoiap committed:
Include tqdm in project dependencies

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

dependabot[bot] committed:
Bump gmsh from 4.11.1 to 4.12.1

Bumps [gmsh](https://gmsh.info) from 4.11.1 to 4.12.1.

---
updated-dependencies:
- dependency-name: gmsh
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump tidy3d from 2.5.1 to 2.5.2

Bumps [tidy3d](https://github.com/flexcompute/tidy3d) from 2.5.1 to 2.5.2.
- [Release notes](https://github.com/flexcompute/tidy3d/releases)
- [Changelog](https://github.com/flexcompute/tidy3d/blob/develop/CHANGELOG.md)
- [Commits](https://github.com/flexcompute/tidy3d/compare/v2.5.1...v2.5.2)

---
updated-dependencies:
- dependency-name: tidy3d
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

joamatab committed:
Merge pull request #295 from sequoiap/patch-1

Include tqdm in project dependencies

joamatab committed:
Merge pull request #298 from gdsfactory/dependabot/pip/tidy3d-2.5.2

Bump tidy3d from 2.5.1 to 2.5.2

joamatab committed:
Merge pull request #287 from gdsfactory/dependabot/github_actions/actions/setup-python-5

Bump actions/setup-python from 4 to 5

joamatab committed:
Merge branch 'main' into dependabot/github_actions/actions/deploy-pages-4

joamatab committed:
Merge pull request #289 from gdsfactory/dependabot/github_actions/actions/deploy-pages-4

Bump actions/deploy-pages from 2 to 4

joamatab committed:
Merge pull request #285 from gdsfactory/dependabot/github_actions/actions/upload-artifact-4

Bump actions/upload-artifact from 3 to 4

joamatab committed:
Merge pull request #288 from gdsfactory/dependabot/github_actions/actions/stale-9

Bump actions/stale from 8 to 9

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.1.11 → v0.1.13](https://github.com/astral-sh/ruff-pre-commit/compare/v0.1.11...v0.1.13)

simbilod committed:
Merge pull request #297 from gdsfactory/dependabot/pip/gmsh-4.12.1

Bump gmsh from 4.11.1 to 4.12.1

joamatab committed:
Merge pull request #302 from gdsfactory/pre-commit-ci-update-config

[pre-commit.ci] pre-commit autoupdate

joamatab committed:
don't report unused expression in pyright

joamatab committed:
fix notebooks not showing

joamatab committed:
revert link

joamatab committed:
Merge pull request #304 from gdsfactory/pyright_remove_unused_expression

Pyright remove unused expression

joamatab committed:
Merge pull request #286 from gdsfactory/dependabot/github_actions/actions/upload-pages-artifact-3

Bump actions/upload-pages-artifact from 2 to 3

joamatab committed:
fix sax notebook

joamatab committed:
Merge pull request #307 from gdsfactory/fix_sax_notebook

fix sax notebook

flaport committed:
bump meow and sax

Commit:
Switch to using a context manager

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

joamatab committed:
Merge pull request #310 from sequoiap/disable_print

Switch to using a context manager for disable_print

simbilod committed:
also ignore layers without thickness or zmin in geometry_layers

joamatab committed:
Merge pull request #311 from simbilod/improve_simulation_component

also ignore layers without thickness or zmin in geometry_layers

dependabot[bot] committed:
Bump gmsh from 4.12.1 to 4.12.2

Bumps [gmsh](https://gmsh.info) from 4.12.1 to 4.12.2.

---
updated-dependencies:
- dependency-name: gmsh
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.1.13 → v0.1.14](https://github.com/astral-sh/ruff-pre-commit/compare/v0.1.13...v0.1.14)

joamatab committed:
Merge pull request #308 from gdsfactory/bump-meow-sax

bump meow and sax

joamatab committed:
Merge pull request #314 from gdsfactory/pre-commit-ci-update-config

[pre-commit.ci] pre-commit autoupdate

Commit:
Update component.py

added ability to user edit "reference_plane" for Tidy3DComponent

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

joamatab committed:
Merge pull request #315 from mkovach2/reference_plane

Update component.py

seanlam-ece committed:
[LUM] Convert layerstack to process file

seanlam-ece committed:
[LUM] Replace geometry generation with Layer Builder version

seanlam-ece committed:
[LUM] Handle dopants gracefully for Lum versions 2021 and 2023

dependabot[bot] committed:
Bump jupyter-book from 0.15.1 to 1.0.0

Bumps [jupyter-book](https://github.com/executablebooks/jupyter-book) from 0.15.1 to 1.0.0.
- [Release notes](https://github.com/executablebooks/jupyter-book/releases)
- [Changelog](https://github.com/executablebooks/jupyter-book/blob/master/CHANGELOG.md)
- [Commits](https://github.com/executablebooks/jupyter-book/compare/v0.15.1...v1.0.0)

---
updated-dependencies:
- dependency-name: jupyter-book
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

joamatab committed:
Merge pull request #316 from gdsfactory/dependabot/pip/jupyter-book-1.0.0

Bump jupyter-book from 0.15.1 to 1.0.0

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate

updates:
- [github.com/PyCQA/bandit: 1.7.6 → 1.7.7](https://github.com/PyCQA/bandit/compare/1.7.6...1.7.7)

joamatab committed:
Merge pull request #317 from gdsfactory/pre-commit-ci-update-config

[pre-commit.ci] pre-commit autoupdate

HelgeGehring committed:
allow multiple mappings for the same layer

HelgeGehring committed:
Merge branch 'main' into component_with_local_layers

HelgeGehring committed:
Merge pull request #318 from HelgeGehring/component_with_local_layers

Component with local layers

simbilod committed:
background meshing

simbilod committed:
Merge branch 'main' into dependabot/pip/gmsh-4.12.2

simbilod committed:
xy file too

simbilod committed:
upgrade meshwell

simbilod committed:
delete scratch folder

simbilod committed:
pass optimization flag

simbilod committed:
upgrade meshwell

joamatab committed:
Merge pull request #320 from simbilod/mesh_smoothing

simbilod committed:
Merge branch 'main' into background_mesh

joamatab committed:
Merge pull request #319 from simbilod/background_mesh

joamatab committed:
Merge branch 'main' into dependabot/pip/gmsh-4.12.2

simbilod committed:
Merge pull request #312 from gdsfactory/dependabot/pip/gmsh-4.12.2

Bump gmsh from 4.12.1 to 4.12.2

dependabot[bot] committed:
Bump actions/cache from 3 to 4

Bumps [actions/cache](https://github.com/actions/cache) from 3 to 4.
- [Release notes](https://github.com/actions/cache/releases)
- [Changelog](https://github.com/actions/cache/blob/main/RELEASES.md)
- [Commits](https://github.com/actions/cache/compare/v3...v4)

---
updated-dependencies:
- dependency-name: actions/cache
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump codecov/codecov-action from 3 to 4

Bumps [codecov/codecov-action](https://github.com/codecov/codecov-action) from 3 to 4.
- [Release notes](https://github.com/codecov/codecov-action/releases)
- [Changelog](https://github.com/codecov/codecov-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/codecov/codecov-action/compare/v3...v4)

---
updated-dependencies:
- dependency-name: codecov/codecov-action
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.1.14 → v0.2.0](https://github.com/astral-sh/ruff-pre-commit/compare/v0.1.14...v0.2.0)
- [github.com/kynan/nbstripout: 0.6.1 → 0.7.1](https://github.com/kynan/nbstripout/compare/0.6.1...0.7.1)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

joamatab committed:
require labels on PRs

joamatab committed:
remove stale workflow

joamatab committed:
Merge pull request #329 from gdsfactory/add_labels

require labels on PRs

joamatab committed:
Bump to 0.9.12

joamatab committed:
update changelog

joamatab committed:
pin latest working version of jax and jaxlib for sax to work

joamatab committed:
Merge pull request #322 from gdsfactory/dependabot/github_actions/codecov/codecov-action-4

Bump codecov/codecov-action from 3 to 4

joamatab committed:
use <= instead of pinning exact version

joamatab committed:
Merge pull request #332 from gdsfactory/pin_jax

pin latest working version of jax and jaxlib for sax to work

joamatab committed:
pin older version of meshwell

joamatab committed:
update meshwell

joamatab committed:
Merge pull request #334 from gdsfactory/pin_meshwell

pin older version of meshwell

joamatab committed:
Bump to 0.9.13

joamatab committed:
update changelog

nikosavola committed:
Mention correct pinned pyvis version

joamatab committed:
Merge pull request #338 from gdsfactory/pinned-pyvis-version-mention

Mention correct pinned pyvis version

joamatab committed:
Merge pull request #324 from gdsfactory/pre-commit-ci-update-config

[pre-commit.ci] pre-commit autoupdate

joamatab committed:
Merge pull request #321 from gdsfactory/dependabot/github_actions/actions/cache-4

Bump actions/cache from 3 to 4

elamdf commented on gplugins/lumerical/tests/test_lumerical_utils.py:
```diff
@@ -0,0 +1,249 @@
+            + "\nThis is used to capture Lumerical versions 2021 that only work with passive only process files "
+            + "and subsequent versions of Lumerical that support both passives and dopants."
+        )
+    print(message)
```
Probably want to use gdsfactory's logging interface just for consistency

elamdf commented on gplugins/lumerical/tests/test_lumerical_utils.py:
```diff
@@ -0,0 +1,249 @@
+    success = False
+    message = ""
+    try:
+        mode = lumapi.MODE(hide=False)
```
I think hide for these should be set by a global/runtime `DEBUG` flag, else this just clutters the desktop for big sweeps. I was using a `DEBUG_LUMERICAL` var, you're welcome to refactor but I think we want to avoid spawning tons of windows that likely won't be looked at with regular usage

elamdf commented on gplugins/lumerical/utils.py:
```diff
@@ -0,0 +1,154 @@
+um = 1e-6
+
+
+def to_lbr(
```
to_lbr would make sense if this were a method of MaterialMap, maybe material_map_to_lbr ?

seanlam-ece replied:
layerstack_to_lbr seems more appropriate since it is more broad. The lbr file encompasses all info about the layerstack.

elamdf commented on gplugins/lumerical/utils.py:
```diff
@@ -0,0 +1,154 @@
+            layer = SubElement(layers, "layer")
+
+            # Default params
+            layer.set("enabled", "1")
```
Maybe worth putting this in a dict somewhere else and just mapping layer.set over the items to improve readability, your call though

elamdf commented on gplugins/lumerical/utils.py:
```diff
@@ -0,0 +1,154 @@
+    with open(str(process_file_path), "w") as f:
+        f.write(xml_str)
+
+    return process_file_path
```
Update doctoring / return type?

elamdf commented on gplugins/lumerical/tests/test_lumerical_utils.py:
```diff
@@ -0,0 +1,249 @@
+                background_doping_ion="Boron",
+                orientation="100",
+            ),
+            # KNOWN ISSUE: Lumerical 2021 version of Layer Builder does not support dopants in process file
```
is there a way we can check the lumerical version (or try/catch somewhere)? being able to draw dopant layers is a pretty important functionality (and the dopant drawings have some extra logic which would benefit from having test coverage)

seanlam-ece replied:
@elamdf We can get the application version (https://optics.ansys.com/hc/en-us/articles/360034932053-version-Script-command), but it does not tell us whether it is from 2021 and prior. The format is like this "7.28.3578". I attempted to look through Lumerical's object API but have yet to find something that may indicate the year in which the software is released. Attempted to look for release notes with these version numbers, but it's not readily apparent...

elamdf replied:
do you think we could try/catch and print a warning if adding dopings to layerstack crash the session? unless lumerical's raised exception is sufficiently verbose

seanlam-ece replied:
The following error message is from Lumerical:
lumapi.LumApiError: 'Please import your process file using the layer builder editor window for more options.'

It's sufficient to warn users of an issue but not specific enough to root cause. Good idea on adding try/catch. I'm going to rework the geometry import to include a better error message.

seanlam-ece replied:
Moved geometry import to a function: https://github.com/seanlam97/gplugins/commit/a9b62d82b43bd3d69df6434ea5b3799143a06a24

elamdf commented on gplugins/lumerical/utils.py:
```diff
@@ -0,0 +1,154 @@
+        This function generates an XML file representing a Layer Builder file for Lumerical, based on the provided the active PDK
+        and material map. It creates 'process.lbr' in the current working directory, containing layer information like name,
+        material, thickness, sidewall angle, and other properties specified in the layer stack. It skips layers that are not
+        of type 'grow' or 'background' and logs a warning for each skipped layer.
```
`doping` and `implant` layers aren't skipped, correct? also nit to make capitalization between the docstring process string comparisons in the code consistent

elamdf replied:
it might be good to have an optional flag to enable/disable dopings (default to enable) for backwards compatibility reasons (otherwise this will always fail being run on lumerical pre-2021R1)

seanlam-ece replied:
> doping and implant layers aren't skipped, correct?

Yes, updated the comments to reflect.

> it might be good to have an optional flag to enable/disable dopings

I like this idea! I've opted for a global var "ENABLE_DOPING"

seanlam-ece replied:
https://github.com/elamdf/gplugins/pull/2/commits/deda87cb2365e8f01c370f61f689f7daf8df58ab

https://github.com/elamdf/gplugins/pull/2/commits/e291774391ab1b32a3042f69b9dfa4a9c967be3c

elamdf commented on gplugins/lumerical/utils.py:
```diff
@@ -0,0 +1,154 @@
+    layers = SubElement(layer_builder, "layers")
+    doping_layers = SubElement(layer_builder, "doping_layers")
+    for layer_name, layer_info in layerstack.to_dict().items():
+        ### Set optical and metal layers
```
not 100% sure what this means- maybe 'add' or 'ingest' is better? also, does doping count as an optical/metal layer? I'm a little confused by the separation here

seanlam-ece replied:
Comment misplaced. Updated. The process file has two sections: one for optical / metal layers and one for doping layers. I suppose they wanted to capture the physical material layers that act as the base stackup then the doping is applied to the layers.

elamdf commented on gplugins/lumerical/utils.py:
```diff
@@ -0,0 +1,154 @@
+            ) and layer_info.get("background_doping_ion", False):
+                doping_layer = SubElement(doping_layers, "layer")
+
+                # Order of param matters
```
nit: params

elamdf commented on gplugins/lumerical/utils.py:
```diff
@@ -0,0 +1,154 @@
+            # the process file correctly and the doping layer will not appear. Therefore, doping layer names MUST be unique.
+            # FIX: Appending "_doping" to name
+
+            # KNOWN ISSUE: If the 'process' is not 'Background' or 'Implant', this will crash CHARGE upon importing process file.
```
is this just for dopings?

seanlam-ece replied:
Yes, doping layers seem to have an issue with the other layers. Will need to report this Lumerical in the future, but this will be our current workaround.

elamdf replied:
I think I understand now- you can only set doping parameters for background/implant layers? that makes sense to me- where's the issue?

seanlam-ece replied:
Line 98: Doping layers in the lbr file can only have process specified as "Background" or "Implant". "Grow" does not make sense for doping layers, so "Grow" is not supported.

If the layer_stack specifies a doping layer as "Grow", when converting to lbr and importing into Layer Builder, the session crashes

elamdf replied:
"grow" shouldn't be a doping layer, right? is your point that lumerical should error (rather than crash) here?

seanlam-ece replied:
Yes. In gdsfactory.generic_tech.LAYER_STACK, the substrate has background_doping_concentration with process set to "grow". When exporting to lbr then importing into Lumerical, Lumerical crashes.

elamdf commented on gplugins/lumerical/write_sparameters_lumerical.py:
```diff
@@ -550,12 +518,120 @@ def _sample_convergence_wavelength():
+        clad="SiO2 (Glass) - Palik",
+    )  # or dict(si=3.45+2j)
+
+    # Create layerstack
```
is this a dummy layersatck? regardless, it should be in its own file- maybe this can be combined with the one used in the test? it's a lot of lines...

seanlam-ece replied:
Yes, it's a dummy layerstack with no doping and fixes the original layerstack in the generic PDK by specifying the clad and box layers as sio2 material with layer_type = "background". It was originally set to "grow", which prevents those layers from being generated properly in Lumerical.

This is all test code anyways so shall we remove it completely?

elamdf replied:
is this the same as `layerstack_lumerical2023` ? I'm not quite sure what this is achieving- to confirm, this whole fie will be removed when this functionality is migrated to the `fdtdDesignRecipe`, right?

seanlam-ece replied:
Agreed, it's test code and can be removed.

> this whole fie will be removed when this functionality is migrated to the fdtdDesignRecipe, right?


Partially yes, the write_sparameters_lumerical is the combination of simulation features (setup, running, extraction) and we will separate this into:
1. fdtd.py
LumericalFdtdSimulation class (simulation setup and data extraction/visualization features)
2. Under lumerical/design_recipe directory
FdtdDesignRecipe (specifies design intent) that invokes functions from LumericalFdtdSimulation and performs additional data processing to form compact models or contains optimizations
3. simulation_settings.py
FdtdSimulationSetup (Pydantic Base Model)
4. convergence_settings.py
FdtdConvergenceSetup (Pydantic Base Model)

elamdf reviewed (commented):
Thanks for doing this! The Layerbuilder interface makes drawing geometry a lot cleaner. Left some (mostly style) comments, but overall looks good. The `write_sparameters_lumerical` functionality (along with its utilization of the lbr) should be migrated to the `fdtdDesignRecipe`, but we can do that in a seperate PR.

elamdf commented on gplugins/lumerical/tests/test_lumerical_utils.py:
```diff
@@ -0,0 +1,249 @@
+        mode.addlayerbuilder()
+        mode.loadprocessfile(str(process_file_lumerical2023))
+        success = success or True
+        message += "\nSUCCESS: Passives and dopants process file successfully loaded."
```
should probably have various handles (MODE, CHARGE, FDTD) and just loop through them, importing the process file

seanlam-ece replied:
Good point! Added

seanlam-ece replied:
https://github.com/elamdf/gplugins/pull/2/commits/6aa4e90717e819aa33dd3cc9f62491182096d8c4

elamdf commented on gplugins/lumerical/tests/test_lumerical_utils.py:
```diff
@@ -0,0 +1,249 @@
+from gdsfactory.technology.layer_stack import LayerStack, LayerLevel
+
+
+def test_to_lbr():
```
I think a short docstring describing what this does (maybe "Ensure the process file generated from `to_lbr` is importable in MODE, FDTD, and CHARGE") would be useful

seanlam-ece replied:
Addressed: https://github.com/elamdf/gplugins/pull/2/commits/13380939a7263ea27afa7a9943002c2b9d494b32

simbilod committed:
basic python-driven density analysis, using klayout tiling processor

simbilod committed:
add test

simbilod committed:
example, fix test

joamatab committed:
Merge pull request #339 from simbilod/density_analytics

basic python-driven density analytics, using klayout tiling processor

simbilod committed:
weighted sum of density for better estimate

simbilod committed:
remove print, plot

joamatab committed:
Merge pull request #340 from simbilod/density_analytics

better global density estimation

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.2.0 → v0.2.1](https://github.com/astral-sh/ruff-pre-commit/compare/v0.2.0...v0.2.1)

seanlam-ece committed:
[LUM] Switch print to logger

seanlam-ece committed:
[LUM] Use global debug var

seanlam-ece committed:
[LUM] Rename to_lbr function

seanlam-ece committed:
[LUM] Combine setting params into dict for better readability; rename to_lbr

seanlam-ece committed:
[LUM] Update lbr return docstring

seanlam-ece committed:
[LUM] Add flag to enable/disable doping layer generation; update docstrings

seanlam-ece committed:
[LUM] Use global var to enable/disable dopings

seanlam-ece committed:
[LUM] Make global var for enabling/disabling doping

seanlam-ece committed:
[LUM] Loop through FDTD, CHARGE, and MODE to test lbr generation; reformat using black

seanlam-ece committed:
[LUM] Add docstring to describe test

seanlam-ece committed:
Format using black

joamatab committed:
better pyproject

joamatab committed:
Merge pull request #341 from gdsfactory/pre-commit-ci-update-config

[pre-commit.ci] pre-commit autoupdate

joamatab committed:
Merge pull request #342 from gdsfactory/better_pyproject

better pyproject

seanlam-ece committed:
[LUM] Make common geometry import function with more detailed error messages

seanlam-ece committed:
Merge branch 'main' of https://github.com/gdsfactory/gplugins

seanlam-ece committed:
[LUM] Reduce test code

seanlam-ece committed:
[LUM] Clarify comment on dopants causing crashes in Lumerical

seanlam-ece committed:
[LUM] Check each simulator for importing process file

seanlam-ece committed:
Fix issues with pre-commit

simbilod committed:
update pyproject

dependabot[bot] committed:
Bump release-drafter/release-drafter from 5 to 6

Bumps [release-drafter/release-drafter](https://github.com/release-drafter/release-drafter) from 5 to 6.
- [Release notes](https://github.com/release-drafter/release-drafter/releases)
- [Commits](https://github.com/release-drafter/release-drafter/compare/v5...v6)

---
updated-dependencies:
- dependency-name: release-drafter/release-drafter
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

simbilod committed:
fix xy meshing

joamatab committed:
Merge pull request #343 from simbilod/parametrize_cellname_density

add cellname parametrization, update pyproject for pre-commit to run

joamatab committed:
Merge pull request #348 from gdsfactory/dependabot/github_actions/release-drafter/release-drafter-6

Bump release-drafter/release-drafter from 5 to 6

joamatab committed:
Merge pull request #349 from simbilod/fix_xy_meshing

fix xy meshing

seanlam-ece committed:
Merge branch 'main' of https://github.com/gdsfactory/gplugins
